cmd/istiod/istiod.go

```diff
@@ -15,144 +15,110 @@
+
+	// Create a test pilot discovery service configured to watch the tempDir.
+	args := &bootstrap.PilotArgs{
+		Config: bootstrap.ConfigArgs{
+			ControllerOptions: kubecontroller.Options{
+				DomainSuffix: "cluster.local",
+				TrustDomain:  "cluster.local",
+			},
+		},
+		Service: bootstrap.ServiceArgs{
+			Registries: []string{string(serviceregistry.Kubernetes)},
+		},
+		InjectionOptions: bootstrap.InjectionOptions{
+			InjectionDirectory: "./var/lib/istio/inject",
+			Port:               basePort + 17,
+		},
+		Mesh: bootstrap.MeshArgs{
+			ConfigFile: "./var/lib/istio/config/mesh",
+		},
+
+		Plugins: bootstrap.DefaultPlugins, // TODO: Should it be in MeshConfig ? Env override until it's done.
+
+		// MCP is messing up with the grpc settings...
+		MCPMaxMessageSize:        1024 * 1024 * 64,
+		MCPInitialWindowSize:     1024 * 1024 * 64,
+		MCPInitialConnWindowSize: 1024 * 1024 * 64,
+		BasePort:                 basePort,
 	}
 
-	// Load the mesh config. Note that the path is slightly changed - attempting to move all istio
-	// related under /var/lib/istio, which is also the home dir of the istio user.
-	istiods, err := istiod.NewIstiod(kcfg, client, "/var/lib/istio/config")
-	if err != nil {
-		log.Fatalf("Failed to start istiod: %v", err)
+	// If the namespace isn't set, try looking it up from the environment.
+	if args.Namespace == "" {
+		args.Namespace = bootstrap.PodNamespaceVar.Get()
 	}
-
-	// Create k8s-signed certificates. This allows injector, validation to work without Citadel, and
-	// allows secure SDS connections to Istiod.
-	initCerts(istiods, client)
-
-	// Init k8s related components, including Galley K8S controllers and
-	// Pilot discovery. Code kept in separate package.
-	k8sServer, err := k8s.InitK8S(istiods, client, kcfg, istiods.Args)
-	if err != nil {
-		log.Fatalf("Failed to start k8s controllers: %v", err)
+	if args.KeepaliveOptions == nil {
+		args.KeepaliveOptions = istiokeepalive.DefaultOption()
 	}
-
-	err = istiods.InitDiscovery()
-	if err != nil {
-		log.Fatalf("Failed to init XDS server: %v", err)
+	if args.Config.ClusterRegistriesNamespace == "" {
+		args.Config.ClusterRegistriesNamespace = args.Namespace
 	}
-
-	if _, err := k8sServer.InitK8SDiscovery(istiods, kcfg, istiods.Args); err != nil {
-		log.Fatalf("Failed to init Kubernetes discovery: %v", err)
+	args.DiscoveryOptions = bootstrap.DiscoveryServiceOptions{
+		GrpcAddr:        fmt.Sprintf(":%d", basePort+10),
+		SecureGrpcAddr:  "",
+		EnableProfiling: true,
+		MonitoringAddr:  fmt.Sprintf(":%d", basePort+18), // TODO: merge all monitoring ports
 	}
-
-	err = istiods.Start(stop, k8sServer.OnXDSStart)
-	if err != nil {
-		log.Fatalf("Failed on start XDS server: %v", err)
+	if basePort == 15000 {
+		args.DiscoveryOptions.HTTPAddr = ":8080" // lots of tools use this port
+	} else {
+		// Running a second Istiod on the same machine, avoid port conflicts.
+		args.DiscoveryOptions.HTTPAddr = fmt.Sprintf(":%d", basePort+80)
```

Comment:
All of the offsets need to be constants (defined in a common place) so that they don't collide.

Reply:
I can remove the standalone istiod component if it avoids the confusion. 

It is still WIP - the intent is to use it as 'only mesh.yaml' and do more risky cleanups.